File: newtonRaphson.R
#.rs.restartR()

rm(list = ls())

setPolynomial <- function() {

  p_ <<- eval(parse(text = readline(prompt = "Enter Polynomial Coefficients: ")))

}; setPolynomial() ## Must be in vector format, e.g. c(1,2,1)

f <- function(x) { ## Horner's rule to evaluate polynomial
  
  y <- p_[1]
  
  for (i in 2:length(p_)) {
    
    y <- y * x + p_[i]
  }
  return(y)
}

fd <- function(x, h = 0.005) { ## Numerical differentiation
  
  return((f(x + h) - f(x - h))/(2 * h))
}

plotf <- function(x = seq(-1.5, 1.5, 0.1)) { ## Get approximate root visually
  
  plot(x, f(x), type = 'l')
  min(f(x))

}; plotf()

setX <- function() { ## Enter initial guess
  
  x_ <<- eval(parse(text = readline(prompt = "Enter x: ")))

}; setX()

newtonRaphson <- function(x0 = x_, tol = 0.0005, lim = 1000) {
  
  for (i in 1:lim) {
    
    x1 <- x0 - f(x0)/fd(x0)
    print(x1)
    
    if (abs(x1 - x0) < tol) {
      
      x0 <<- x0
      x1 <<- x1
      break;
    }
    x0  <- x1
  }
}; newtonRaphson();

File: newtonRaphson.cpp
#include<bits/stdc++.h>

using namespace std;

class impliedVol {

	int len; 
	int lim;
	double c;
	double x;
	double tol;
	
	vector<double> p;
	
	public:
	
		void setP() { // Set polynomial coefficients, degree n to 1
			
			cout << "Set vector length: " << endl;
			cin >> len;
			
			for (int i = 1; i <= len; i++) {
				
				cout << "Set " << i << "th element: " << endl;
				cin >> c;
				
				p.push_back(c);
			}
		}		
		void setInputs() { // Set inputs
			
			cout << "Set x for f(x): " << endl;
			cin >> x;
			
			cout << "Set tol for abs(x1-x0) < tol: " << endl;
			cin >> tol;

			cout << "Set lim for iteration limit: " << endl;
			cin >> lim;
		}
		
		double f(double x) { // Horner's Rule to evaluate polynomial
			
			double y = p[0];
			
			for (int i = 1; i <= p.size(); i++) {
				
				y = y * x + p[i];
			}
			return y;
		}
		
		double fd(double x, double h = 0.005) { // Numerical differentiation
				
			double yd;
					
			yd = (f(x + h) - f(x - h))/(2 * h);
			
			return yd;
		}
		
		double newtonRaphson() {
			
			double x0 = x;
			double x1;
			
			int i = 1;

			while (i < lim) {

				x1 = x0 - f(x0)/fd(x0);
				
				cout << x1 << endl;
				
				if (abs(x1 - x0) <= tol) { 
				
					return(x1);
				}
				x0 = x1;
			}
			return(9999);
		}
};

int main() {
	
	impliedVol* iv = new impliedVol();
	
	iv->setP();
	iv->setInputs();
	iv->newtonRaphson();
	
	delete iv;
}
